sklearn/metrics/_pairwise_distances_reduction/_radius_neighbors_classmode.pyx.tp

```diff
@@ -0,0 +1,220 @@
+        self.outlier_label_exists = False
+        self.outlier_label = outlier_label
+
+        cdef intp_t idx
+        if outlier_label is not None:
+            self.outlier_label_exists = True
+            for idx in range(self.unique_labels.shape[0]):
+                if self.unique_labels[idx] == outlier_label:
+                    self.outlier_index = idx
+
+        # Map from set of unique labels to their indices in `class_scores`
+        # Buffer used in building a histogram for one-pass weighted mode
+        self.class_scores = np.zeros(
+            (self.n_samples_X, unique_labels.shape[0]), dtype=np.float64,
+        )
+
+
+    cdef inline void weighted_histogram_mode(
+        self,
+        intp_t sample_index,
+        intp_t k,
+        intp_t* indices,
+        float64_t* distances,
+    ) noexcept nogil:
+        cdef:
+            intp_t neighbor_idx, neighbor_class_idx, label_index
+            float64_t score_incr = 1
+            bint use_distance_weighting = (
+                self.weight_type == WeightingStrategy.distance
+            )
+
+        if k == 0:
+            self.outliers.push_back(sample_index)
+            if self.outlier_index >= 0:
+                self.class_scores[sample_index][self.outlier_index] = score_incr
+
+            return
+
+        # Iterate over the neighbors. This can be different for
+        # each of the samples as they are based on the radius.
+        for neighbor_rank in range(k):
+            if use_distance_weighting:
+                score_incr = 1 / distances[neighbor_rank]
+
+            neighbor_idx = indices[neighbor_rank]
+            neighbor_class_idx = self.class_membership[neighbor_idx]
+            self.class_scores[sample_index][neighbor_class_idx] += score_incr
+
+        return
+
+    @final
+    cdef void _parallel_on_X_prange_iter_finalize(
+        self,
+        intp_t thread_num,
+        intp_t X_start,
+        intp_t X_end,
+    ) noexcept nogil:
+        cdef:
+            intp_t idx
+
+        for idx in range(X_start, X_end):
+            self.weighted_histogram_mode(
+                sample_index=idx,
+                k=deref(self.neigh_indices)[idx].size(),
+                indices=deref(self.neigh_indices)[idx].data(),
+                distances=deref(self.neigh_distances)[idx].data(),
+            )
+
+        return
+
+    cdef void _parallel_on_Y_finalize(
+        self,
+    ) noexcept nogil:
+        cdef:
+            intp_t idx
+
+        with nogil, parallel(num_threads=self.effective_n_threads):
+            # Merge vectors used in threads into the main ones.
+            # This is done in parallel sample-wise (no need for locks).
+            for idx in prange(self.n_samples_X, schedule='static'):
+                self._merge_vectors(idx, self.chunks_n_threads)
+
+            for idx in prange(self.n_samples_X, schedule='static'):
+                self.weighted_histogram_mode(
+                    sample_index=idx,
```

Comment:
@jjerphan I might need some guidance here. This is failing when parallel on Y is used giving errors like
```console
python(24844,0x16ef6f000) malloc: double free for ptr 0x11cf39200
python(24844,0x16f777000) malloc: double free for ptr 0x11cf36600
python(24844,0x16e767000) malloc: double free for ptr 0x11cf39200
python(24844,0x16df5f000) malloc: Double free of object 0x11c27d010
python(24844,0x16df5f000) malloc: *** set a breakpoint in malloc_error_break to debug
python(24844,0x16ef6f000) malloc: *** set a breakpoint in malloc_error_break to debug
python(24844,0x16e767000) malloc: *** set a breakpoint in malloc_error_break to debug
python(24844,0x16f777000) malloc: *** set a breakpoint in malloc_error_break to debug
```
I think I might not be correctly accessing the samples in this case. Could you kindly check?

Reply:
I think this is because of the prange. If I remove prange it works fine.